File: utilities/shell/direnv-config/sdk/elixir/lib/direnv_config/resolve.ex
defmodule DirenvConfig.Resolve do
  @moduledoc """
  Resolves the `.active` configuration by merging YAML layers.

  Layer merge order: base.yaml -> {DC_ENV}.yaml -> local.yaml -> secrets.yaml
  """

  @spec resolve_active(String.t(), String.t()) :: {:ok, term()} | {:error, term()}
  def resolve_active(store_path, name) do
    config_dir = Path.join(store_path, name)
    env = System.get_env("DC_ENV") || "dev"

    layer_names = ["base", env, "local", "secrets"]

    layers =
      layer_names
      |> Enum.map(fn layer -> Path.join(config_dir, "#{layer}.yaml") end)
      |> Enum.filter(&File.exists?/1)
      |> Enum.reduce([], fn file, acc ->
        case YamlElixir.read_from_file(file) do
          {:ok, data} when is_map(data) -> acc ++ [data]
          _ -> acc
        end
      end)

    merged =
      case layers do
        [] -> %{}
        _ -> DirenvConfig.Merge.deep_merge_multi(layers) || %{}
      end

    File.mkdir_p!(config_dir)
    active_path = Path.join(config_dir, ".active")
    yaml_content = Ymlr.document!(merged)
    File.write!(active_path, yaml_content)

    {:ok, merged}
  rescue
    e -> {:error, e}
  end
end

File: utilities/shell/direnv-config/sdk/typescript/src/resolve.ts
import { readFile, writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import { deepMergeMulti } from './merge.js';

export async function resolveActive(storePath: string, name: string): Promise<unknown> {
  const env = process.env['DC_ENV'] ?? 'dev';
  const layerNames = ['base', env, 'local', 'secrets'];

  const layers: unknown[] = [];
  for (const layer of layerNames) {
    const filePath = join(storePath, name, `${layer}.yaml`);
    if (existsSync(filePath)) {
      const content = await readFile(filePath, 'utf-8');
      const parsed = parseYaml(content);
      if (parsed !== null && parsed !== undefined) {
        layers.push(parsed);
      }
    }
  }

  const merged = deepMergeMulti(layers);
  const activePath = join(storePath, name, '.active');
  await writeFile(activePath, stringifyYaml(merged), 'utf-8');
  return merged;
}
